File: Session_01_AlgorithmAnalysis/multi_language_implementations/complexity_examples.js
const readline = require('readline');

function demonstrateConstantTime(n) {
    let operations = 0;
    let a = 10;
    operations++;
    let b = 20;
    operations++;
    let sum = a + b;
    operations++;
    console.log("O(1) Constant Time:");
    console.log(`  Input size (n): ${n}`);
    console.log(`  Actual operations counted: ${operations}\n`);
}

function demonstrateLogarithmicTime(n) {
    let operations = 0;
    let temp = n;
    console.log("O(log n) Logarithmic Time:");
    console.log(`  Dividing input size ${n} by 2 repeatedly:`);
    while (temp > 1) {
        console.log(`    Current value: ${temp}`);
        temp = Math.floor(temp / 2);
        operations++;
    }
    console.log(`  Input size (n): ${n}`);
    console.log(`  Actual operations counted: ${operations}\n`);
}

function demonstrateLinearTime(n) {
    let operations = 0;
    for (let i = 0; i < n; i++) {
        operations++;
    }
    console.log("O(n) Linear Time:");
    console.log(`  Input size (n): ${n}`);
    console.log(`  Actual operations counted: ${operations}\n`);
}

function demonstrateLinearithmicTime(n) {
    let operations = 0;
    for (let i = 0; i < n; i++) {
        let temp = n;
        while (temp > 1) {
            temp = Math.floor(temp / 2);
            operations++;
        }
    }
    console.log("O(n log n) Linearithmic Time:");
    console.log(`  Input size (n): ${n}`);
    console.log(`  Actual operations counted: ${operations}\n`);
}

function demonstrateQuadraticTime(n) {
    let operations = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            operations++;
        }
    }
    console.log("O(n^2) Quadratic Time:");
    console.log(`  Input size (n): ${n}`);
    console.log(`  Actual operations counted: ${operations} (n * n)\n`);
}

function main() {
    let n = 16;
    console.log("=== Time Complexity Visualizer ===");
    console.log("This program counts operations to demonstrate theoretical time complexities.\n");
    demonstrateConstantTime(n);
    demonstrateLogarithmicTime(n);
    demonstrateLinearTime(n);
    demonstrateLinearithmicTime(n);
    demonstrateQuadraticTime(n);
    
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();

File: Session_01_AlgorithmAnalysis/multi_language_implementations/complexity_examples.swift
import Foundation

func demonstrateConstantTime(_ n: Int) {
    var operations = 0
    let a = 10
    operations += 1
    let b = 20
    operations += 1
    let _ = a + b
    operations += 1
    print("O(1) Constant Time:")
    print("  Input size (n): \(n)")
    print("  Actual operations counted: \(operations)\n")
}

func demonstrateLogarithmicTime(_ n: Int) {
    var operations = 0
    var temp = n
    print("O(log n) Logarithmic Time:")
    print("  Dividing input size \(n) by 2 repeatedly:")
    while temp > 1 {
        print("    Current value: \(temp)")
        temp /= 2
        operations += 1
    }
    print("  Input size (n): \(n)")
    print("  Actual operations counted: \(operations)\n")
}

func demonstrateLinearTime(_ n: Int) {
    var operations = 0
    for _ in 0..<n {
        operations += 1
    }
    print("O(n) Linear Time:")
    print("  Input size (n): \(n)")
    print("  Actual operations counted: \(operations)\n")
}

func demonstrateLinearithmicTime(_ n: Int) {
    var operations = 0
    for _ in 0..<n {
        var temp = n
        while temp > 1 {
            temp /= 2
            operations += 1
        }
    }
    print("O(n log n) Linearithmic Time:")
    print("  Input size (n): \(n)")
    print("  Actual operations counted: \(operations)\n")
}

func demonstrateQuadraticTime(_ n: Int) {
    var operations = 0
    for _ in 0..<n {
        for _ in 0..<n {
            operations += 1
        }
    }
    print("O(n^2) Quadratic Time:")
    print("  Input size (n): \(n)")
    print("  Actual operations counted: \(operations) (n * n)\n")
}

func main() {
    let n = 16
    print("=== Time Complexity Visualizer ===")
    print("This program counts operations to demonstrate theoretical time complexities.\n")
    demonstrateConstantTime(n)
    demonstrateLogarithmicTime(n)
    demonstrateLinearTime(n)
    demonstrateLinearithmicTime(n)
    demonstrateQuadraticTime(n)
    print("Press Enter to exit...", terminator: "")
    fflush(stdout)
    let _ = readLine()
}

main()
